# Data Mining Versuch Music Clustering
* Autor: Prof. Dr. Johannes Maucher
* Datum: 16.10.2015

[Übersicht Ipython Notebooks im Data Mining Praktikum](Data Mining Praktikum.ipynb)

# Einführung
## Lernziele:
In diesem Versuch sollen Kenntnisse in folgenden Themen vermittelt werden:

* Zugriff auf Musikdateien
* Transcodierung von mp3 zu wav 
* Extraktion von Merkmalen in Musikdateien (Feature Extraction)
* Optimierung mit dem genetischen Algorithmus
* Selektion der aussagekräftigsten Merkmale (Feature Selection)
* Clustering von Musikfiles (automatische Playlistgenerierung)


## Vor dem Versuch zu klärende Fragen

### Transcodierung von MP3 nach WAV und Merkmalsextraktion
In diesem Versuch wird der MP3 Decoder [mpg123](http://www.mpg123.de/) eingesetzt. Installieren und testen sie diesen Decoder vor dem Versuch auf ihrem Rechner. Machen Sie sich zunächst mit dem in Kapitel [Gegebene Module zur Transcodierung und Feature Extraction](#Gegebene-Module-zur-Transcodierung-und-Feature-Extraction) aufgeführten Code vertraut. Versuchen Sie Funktion und Ablauf dieses Programms zu verstehen und beantworten Sie folgende Fragen.





1. Was versteht man unter den statistischen Größen _Mittelwert, Standardabweichung, Skewness und Kurtosis_?

#### Mittelwert
Mit dem Begriff Mittelwert (_auch Durchschnitt genannt_) ist in der Statistik meistens das _artithmetische Mittel_ gemeint. Der Mittelwert ist dabei eine Kennzahl für die zentrale Tendenz einer Verteilung. Das arithmetische Mittel ist rechnerisch die Summe der gegebenen Werte geteilt durch die Anzahl der Werte. Angewandt kann dies beispielsweise zum Berechnen einer Durchschnittsgeschwindigkeit werden. In diesem Fall würden die Werte als Geschwindigkeit interpretiert werden. Der _Erwartungswert_ einer Zufallsvariablen beschreibt hingegen die Zahl, die die Zufallsvariable im Mittel annimmt. 

#### Standardabweichung 
Die Standardabweichung ist ein Maß dafür, wie weit einzelne Datensätze verteilt sind. Mathematisch wird die Standardabweichung definiert als die mittlere quadratische Abweichung einer reellen Zufallsvariablen von ihrem Erwartungswert. Wenn alle Werte gleich sind, so ist die Standardabweichung Null, da diese von der Varianz abgeleitet ist. Das heißt je weiter Daten vertreut sind, desto höher ist die Standardabweichung. Je enger/näher Datensätze jedoch beieinander liegen, desto niedriger fällt die Standardabweichung aus. Es gibt zwei verschiedene Formeln der Anwendung, welche sich darin unterscheiden, dass nicht durch _n_, sondern durch _n-1_ geteilt wird.
- die Standardabweichung für eine Stichprobe, wenn die Ergebnisse generalisierbar sein sollen. D.h. man möchte Ergebnisse erzielen, die es ermöglichen auf alle Menschen und nicht nur auf die ursprügnliche Versuchsgruppe _x_ Rückschlüsse zu ziehen.
- die Standardabweichung für die Grundgesamtheit, wenn die Ergebnisse nicht verallgemeinert werden sollen. D.h. die Ergebnisse gelten nur für die Versuchsgruppe _x_.

#### Skewness
Die Maßzahl der Schiefe (_eng. skewness_) gibt an, inwieweit eine Verteilungsfunktion sich zu einer Seite "neigt". Das heißt diese ist ein Maß für die Asymmetrie einer Verteilung. Der Wert kann dabei _positiv_ (Verteilungsfunktion tendiert nach rechts), _negativ_ (Verteilungsfunktion tendiert nach links), _null_ (Verteilungsfunktion ist symmetrisch) und _undefiniert_ (0/0) sein.

###### Eigenschaften einer unimodalen Verteilung**
- linksschief (identisch mit dem Begriff rechtssteil) Verteilungen ist der Median größer als das arithmetische Mittel. 
- rechtsschief  (identisch mit dem Begriff linkssteil) Verteilungen ist der Modus kleiner als der Erwartungswert
- Ist die Verteilungsfunktion symmetrisch, so ist das arithmetische Mittel gleich dem Median und die Verteilung wird eine Schiefe von 0 haben. 

#### Kurtosis
Die Kurtosis (_dt. Wölbung_) ist ein Maß für die Steilheit einer Wahrscheinlichkeitsfunktion, oder Häufigkeitsverteilung. Verteilungen mit geringer Wölbung streuen relativ gleichmäßig.Verteilungen mit niederger Kurtosis streuen relativ gleichmäßig(breite Glocke), Verteilungen mit hoher Kurtosis haben eine mehr extreme auf einem Punkt verteilte Streuung(spitze Glocke).

2. Was beschreibt die Fourier-Transformierte eines zeitlich ausgedehnten Signals?

#### Die Fourier Transformation
Die Fourier-Transformation (genauer die kontinuierliche Fourier-Transformation  ist eine mathematische Methode aus dem Bereich der Fourier-Analysis, mit der kontinuierliche, aperiodische Signale in ein kontinuierliches Spektrum zerlegt werden. Die Funktion, die dieses Spektrum beschreibt, nennt man auch Fourier-Transformierte oder Spektralfunktion. Es handelt sich dabei um eine Integraltransformation, die nach dem Mathematiker Jean Baptiste Joseph Fourier benannt ist. Fourier führte im Jahr 1822 die Fourier-Reihe ein, die jedoch nur für periodische Signale definiert ist und zu einem diskreten Frequenzspektrum führt. Die Entwicklung einer Funktion in ihre _Fourier-Reihe_ wird harmonische Analyse genannt. Die Funktion wird dabei als Überlagerung von sinusförmigen, Schwingungen dargestellt. Ist die Funktion ein Eingangssignal eines LTI-Systems, kann das Ausgangssignal relativ einfach berechnet werden, da das Signals als Überlagerung von Eigenfunktionen des Systems vorliegt.

Kurz: Die _Fourier Transformation_ zerlegt das Signal in verschiedene Sinuswellen, aus denen es besteht. Sie gibt an mit welcher Intensität die einzelnen Frequenzen im zeitlich ausgedehnten Signal vorkommen.

3. Mit welcher Samplingrate werden die WAV Dateien abgetastet?

#### Abtastrate von WAV Dateien
Die Abtastrate oder Abtastfrequenz, auch Samplingrate, Samplerate oder Samplingfrequenz, ist in der Signalverarbeitung die Häufigkeit, mit der ein Analogsignal (auch zeitkontinuierliches Signal genannt) in einer vorgegebenen Zeit abgetastet (das heißt, gemessen und in ein zeitdiskretes Signal umgewandelt) wird. Da es sich bei einer Abtastung immer um einen periodischen Vorgang handelt, ist ihre grundlegende Einheit das *Hertz* (abgekürzt: Hz), also Perioden pro Sekunde.

- Ein Abtastvorgang pro Sekunde: 1 Hz = 1 S/s
- Bei CDs werden die Audiodateien als WAV-Dateien mit 44.1Khz Abtastrate abgespeichert.
- Die Sample Rate dieses Versuches wird in -r 10.000 angegeben.

Die Waveform Audi File Format (_WAV_) ist eine von IBM und Missrn
soft entwickeltes Containerformat für die Speicherung von Audiodaten. Die in den Containern enthaltenen  Dateien sind normalerweise unkomprimierte in _Pulscodemodulation (PCM)_ codierte Audiosignale für die Speicherung und Bearbeitung von Audio-Informationen. WAV-Dateien sind unkomprimiert und fehlerfrei, aber relativ groß. Ihre Größe ergibt sich aus der Abtastrate und der Samplingtiefe bei der Digitalisierung des analogen Audiosignals. Daraus ergibt sich bei einer standardmäßigen Abtastrate von **44,1 kHz** und einer Samplingtiefe von **16 Bit** eine Dateigröße von **5,3 Megabyte (MB) pro Minute** für ein Monosignal. Es sind allerdings auch andere Abtastraten und Sampletiefen möglich.



4. Insgesamt werden 42 Merkmale pro Musiksequenz extrahiert. Beschreiben Sie kurz diese Merkmale

#### Merkmalsextraktion bei WAV Dateien
Die extrahierten Merkmale betreffen die Kurtuosis... **(muss noch ergänzt werden)** Diese Merkmale liegen sowohl im Spektralbereich, als auch Merkmale im Zeitbereich.

_amp1/amp10/amp100/amp1000: Verschieden starke Glättung von 1, 10, 100 und 1000 samples_

- mean: Mittelwert

- dstd: Wir haben dazu nichts gefunden

- std: Standartabweichung

- skew: Skewness

- kurt: Kurtosis

- power1-power10: Spalten den Frequenzbereich in 10 Bereiche und geben den Pegel in dem jeweiligen Bereich an.


### Matching der Teilsequenzen

1. Nachdem für jedes Musikstück die beiden Teilsequenzen in Form der extrahierten Merkmale vorliegen: Wie kann die Ähnlichkeit zwischen Teilsequenzen ermittelt werden?

Die Musikstücke, welche als Trainingsdaten vorliegen, werden die einzelnen Teilsequencen in Vektoren umgewandelt und anschließend auf deren Ähnlichkeit miteinander vergliechen. Haben diese eine hohe Ähnlichkeit, so so müsste das erste Teilstück und das zweite Teilstück gleich oder sehr ähnlich sein. 



2. Welche Numpy- bzw. Scipy-Module können Sie für die Bestimmung der Ähnlichkeit zwischen Teilsequenzen einsetzen?

Aus der Scipy Libary können folgende Funktionen angewandt werden, um die Ähnlichkeit der einzelnen Teilsequenzen zu berechnen.  

```
scipy.spatial.distance.euclidean (Euklid)

scipy.stats.pearsonr (Pearson)

scipy.spatial.distance.cosine (Kosinus)
```


## Genetischer Algorithmus für die Merkmalsselektion

1. Beschreiben Sie die Prozesschritte im genetischen Algorithmus [Genetischer Algorithmus](https://www.hdm-stuttgart.de/~maucher/Python/FunktionenAlgorithmen/html/genAlgTSP.html)



Ein Genetischer Algorithmus (GA) ist eine Such-Heuristik, welche durch _Charls Darwin's_ Theorie der natürlichen Evolution inspiriert wurde. Dieser reflektiert den Prozess, in welchem eine natürliche Selektion der besten (_fittest_) Individuen einer Population für eine Reproduktion selektiert werden. Genetische Algorithmen (GA) sind sehr gut für Probleme in der Suche, als auch für Optimierungen einzusetzen. Ein Beispiel hierfür ist der Einsatz eines _GA_, um eine Lösung für das "Travelling Salesman Problem" (TSP) zu finden. 
Für die Erkundung der besten Merkmale in diesem Fall werden die einzelnen Elemente des GA auf die Problemstellung wie folgt übertragen: 


* **Gene**: Satz an Parametern (Variablen).
* **Individual/Chromosome**: Ein Chromosom ist eine Zusammensetzung von Genen. In diesem Fall ein einzelnes Merkmal, welches die Bedingungen erfüllt. 
* **Population**: Eine Sammlung möglicher Merkmale.
* **Parents**: Zwei Merkmale, welche kombiniert werden, um ein neues Merkmal zu generieren.
* **Mating Pool**: Eine Sammlung an Elternteilen, welche dazu verwendet werden, eine neue Population (nächste Generation) zu generieren.
* **Fitness**: Eine Funktion, welche die Güte der Mermale anhand ihres _mittleren Rangs_ bewertet.
* **Mutation**: Eine Möglichkeit Variationen in der Population zu erzeugen, durch zufälliges austauschen von Elementen der Merkmale.

#### Die Schritte der Ausführung des Genetischen Algorithmus

##### Erzeugung zufälliger Population:
Erzeugt eine Menge von Individuen mit zufälligen Chromosomen. Jedes Individum ist ein Lösungsvorschlag für ein Problem.

##### Bestimme Fitness:
Individuen werden anhand einer Fitnessfunktion bewertet.

##### Selektion:
Wähle ein Paar von Individuen aus. Die Auswahl ist von der Fitness abhängig. Je höher die Fitness, desto höher die Wahrscheinlichkeit, dass sie für die Kreuzung ausgewählt werden.

##### Kreuzung:

Paare werden gekreuzt und erzeugen Nachkommen, welche die Chromosomen der Eltern erben. Hier wird ein zufälliger Schnittpunkt zur Vereinigung der Elternvektoren genutzt. Die Chromosome werden an dieser Stelle geteilt und miteinander vertauscht.

##### Mutation:

Ein Teil der Chromosomen wird zufällig verändert. Es gibt eine Mutationswahrscheinlichkeit, folglich kommt es nicht immer zu einer Mutation.

##### Austausch:

Wenn einer der Nachkommen eine bessere Fitness besitzt als einer aus der Ursprungspopulation, wird dieser ersetzt. Dabei dürfen keine Duplikate entstehen.

##### Abbruchbedingung:

Gibt an wann die Schleife beendet wird. In der Regel wird der genetische Algorithmus beendet wenn nach vielen Iteration die Fitness sich nicht verbessert, manchmal auch nach einer festgelegten Anzahl an Iterationen.


2. In diesem Versuch wird davon ausgegangen, dass Merkmale dann gut sind, wenn durch sie die erste Teilsequenz eines Musikstücks durch einen ähnlichen Vektor wie die jeweils zweite Teilsequenz beschrieben wird. Wie kann mit dieser Annahme der genetische Algorithmus für die Merkmalsselektion angewandt werden. Unter Merkmalsselektion versteht man allgemein die Suche nach den $r$ besten Merkmalen aus einer Menge von insgesamt $R$ Merkmalen. In diesem Versuch werden initial $R=42$ Merkmale extrahiert, aus denen dann die besten $r<R$ Merkmale zu bestimmen sind. Überlegen Sie hierfür speziell wie die Fitnessfunktion, die Kreuzung und die Mutation zu realisieren sind.

##### Fitnessfunktion:

Euklidische Distanz der verwendeten Features zwischen der ersten und der zweiten Teilsequenz. Diese errechneten Werte werden aufsteigend sortiert in einem Array abgespeichert. Danach wird aus dem Array ein Mittlerer Rang(siehe Aufgabebeschreibung Matching der Teilsequenzen 4.) berechnet. Dieser spiegelt die Fitness wieder.

##### Kreuzung:

Indivduen mit niedriger Distanz bzw. mit gutem Fitnesswert werden miteinander gekreuzt. Dabei wird ein zufälliger Schnittpunkt zur Vereinigung der Elternvektoren genutzt. Die Chromosome werden an dieser Stelle geteilt und miteinander vertauscht.

##### Mutation:

Mit niedriger Wahrscheinlichkeit wird ein Wert zufällig verändert.


### Clustering und Playlistgenerierung

1. Wie kann mit einem hierarchischen Clustering der Musikfiles eine Menge von Playlists erzeugt werden, so dass innerhalb einer Playlist möglichst ähnliche Titel zu finden sind?

Titel, die im Clustering nahe beieinander liegen, werden in eine Playlist genommen. Je größer man die Cluster macht, desto unähnlicher werden die Titel untereinander sein. In dem Prozess werden ähnliche Songs zu einem Cluster zusammengefügt. Danach werden Cluster, die sich ähnlich sind, zu Clustern zusammengefügt usw.


# Durchführung
## Gegebene Module zur Transcodierung und Feature Extraction
Mit dem in diesem Abschnitt gegebenen Code werden die im Unterverzeichnis _BandCollection_ befindlichen mp3-Files zunächst in wave decodiert. Danach werden aus den wave Dateien Audiomerkmale erhoben.

Von jedem Musikstück werden zwei disjunkte Teilsequenzen erhoben und von beiden Teilsequenzen jeweils ein Merkmalsvektor gebildet. Der Grund hierfür ist: Für die später folgende Bestimmung der wichtigsten Merkmale (Merkmalsselektion mit dem genetischen Algorithmus), wird angenommen dass Merkmale dann gut sind, wenn die aus ihnen gebildeten Merkmalsvektoren für Teilsequenzen des gleichen Musikstücks nahe beieinander liegen und die Merkmalsvektoren von Teilsequenzen unterschiedlicher Musikstücke weiter voneinander entfernt sind. In der Merkmalsselektion werden dann die Merkmale als relevant erachtet, für die diese Annahme zutrifft. 

**Aufgaben:**

1. Stellen Sie im unten gegebenen Code die Verzeichnisse für Ihre Musikdateien (aktuell Unterverzeichnis _BandCollection_) und für den Ort Ihres _mpg123_ Decoders richtig ein.
2. Die verwendete Musiksammlung sollte mindestens 5 verschiedene Interpreten möglichst unterschiedlicher Genres enthalten. Von jedem Interpret sollten mehrere Titel (evtl. ein ganzes Album) enthalten sein.
3. Führen Sie den in diesem Abschnitt gegebenen Programmcode zur Audiofeature-Extraction aus. Damit werden für alle Musiksequenzen jeweils 42 Merkmale extrahiert. Die extrahierten Merkmalsvektoren der jeweils ersten Sequenz werden in das File _FeatureFileTrainingAllList1.csv_ geschrieben, die der zweiten Teilsequen in das File _FeatureFileTestAllList2.csv_. 


In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
 cd '/content/drive/My Drive/Colab Notebooks'

/content/drive/.shortcut-targets-by-id/1yHcauYAFR9wD5O_2_jLUbkgxOwcTrIcc/Colab Notebooks


In [0]:
# Check for TPU Device with Colab

import os
import pprint
import tensorflow as tf

if 'COLAB_TPU_ADDR' not in os.environ:
  print('ERROR: Not connected to a TPU runtime; please see the first cell in this notebook for instructions!')
else:
  tpu_address = 'grpc://' + os.environ['COLAB_TPU_ADDR']
  print ('TPU address is', tpu_address)

  with tf.Session(tpu_address) as session:
    devices = session.list_devices()
    
  print('TPU devices:')
  pprint.pprint(devices)

In [0]:
import subprocess
import wave
import struct
import numpy
import os
import pandas as pd

numpy.set_printoptions(precision=2,suppress=True)

#Names of features extracted in this module
FeatNames=["amp1mean","amp1std","amp1skew","amp1kurt","amp1dmean","amp1dstd","amp1dskew","amp1dkurt","amp10mean","amp10std",
           "amp10skew","amp10kurt","amp10dmean","amp10dstd","amp10dskew","amp10dkurt","amp100mean","amp100std","amp100skew",
           "amp100kurt","amp100dmean","amp100dstd","amp100dskew","amp100dkurt","amp1000mean","amp1000std","amp1000skew",
           "amp1000kurt","amp1000dmean","amp1000dstd","amp1000dskew","amp1000dkurt","power1","power2","power3","power4",
           "power5","power6","power7","power8","power9","power10"]

In [0]:
def moments(x):
    mean = x.mean()
    std = x.var()**0.5
    skewness = ((x - mean)**3).mean() / std**3
    kurtosis = ((x - mean)**4).mean() / std**4
    return [mean, std, skewness, kurtosis]

In [0]:
#Feature category 2: Frequency domain parameters
def fftfeatures(wavdata):
    f = numpy.fft.fft(wavdata)
    f = f[2:int(f.size / 2 + 1)]
    f = abs(f)
    total_power = f.sum()
    f = numpy.array_split(f, 10)
    return [e.sum() / total_power for e in f]

In [0]:
#Creating the entire feature vector per music-file
def features(x):
    x = numpy.array(x)
    f = []

    xs = x
    diff = xs[1:] - xs[:-1]
    f.extend(moments(xs))
    f.extend(moments(diff))

    xs = x.reshape(-1, 10).mean(1)
    diff = xs[1:] - xs[:-1]
    f.extend(moments(xs))
    f.extend(moments(diff))

    xs = x.reshape(-1, 100).mean(1)
    diff = xs[1:] - xs[:-1]
    f.extend(moments(xs))
    f.extend(moments(diff))

    xs = x.reshape(-1, 1000).mean(1)
    diff = xs[1:] - xs[:-1]
    f.extend(moments(xs))
    f.extend(moments(diff))

    f.extend(fftfeatures(x))
    return f

In [0]:
def read_wav(wav_file):
    """Returns two chunks of sound data from wave file."""
    w = wave.open(wav_file)
    n = 50 * 10000
    if w.getnframes() < n * 3:
        raise ValueError('Wave file too short')
    #For each music file 2 sequences, each containing n frames are subtracted. The first sequence starts at postion n,
    #the second sequence starts at postion 2n. The reason for extracting 2 subsequences is, that later on we like to
    #find the best features and in this exercise we assume that good features have the property that they are similar for 2 subsequences
    #of the same song, but differ for subsequences of different songs.
    w.setpos(n)
    frames = w.readframes(n)
    wav_data1 = struct.unpack('%dh' % n, frames)
    frames = w.readframes(n)
    wav_data2 = struct.unpack('%dh' % n, frames)
    return wav_data1, wav_data2

In [0]:
def compute_chunk_features(mp3_file):
    """Return feature vectors for two chunks of an MP3 file."""
    # Extract MP3 file to a mono, 10kHz WAV file
    #mpg123_command = 'C:\Program Files (x86)\mpg123-1.22.0-x86\mpg123-1.22.0-x86\\mpg123.exe -w "%s" -r 10000 -m "%s"'
    #mpg123_command = 'C:\\Program Files (x86)\\mpg123-1.21.0-x86-64\\mpg123.exe -w "%s" -r 10000 -m "%s"'
    mpg123_command = '/content/drive/My Drive/Colab Notebooks/mpg123/mpg123.exe -w "%s" -r 10000 -m "%s"'
    #mpg123_command = '/content/drive/My Drive/Colab Notebooks/BandCollection
    out_file = 'temp.wav'
    cmd = mpg123_command % (out_file, mp3_file)
    temp = subprocess.call(cmd)
    # Read in chunks of data from WAV file
    wav_data1, wav_data2 = read_wav(out_file)
    # We'll cover how the features are computed in the next section!
    return numpy.array(features(wav_data1)), numpy.array(features(wav_data2))

In [97]:
fileList=[]
featureList1=[]
featureList2=[]
#Specify the name of the directory, which contains your MP3 files here.
# This directory should contain for each band/author one subdirectory, which contains all songs of this author
for path, dirs, files in os.walk('./BandCollection'):
    for f in files:
        if not f.endswith('.mp3'):
            # Skip any non-MP3 files
            continue
        mp3_file = os.path.join(path, f)
        print mp3_file
        # Extract the track name (i.e. the file name) plus the names
        # of the two preceding directories. This will be useful
        # later for plotting.
        tail, track = os.path.split(mp3_file)
        tail, dir1 = os.path.split(tail)
        tail, dir2 = os.path.split(tail)
        # Compute features. feature_vec1 and feature_vec2 are lists of floating
        # point numbers representing the statistical features we have extracted
        # from the raw sound data.
        try:
            feature_vec1, feature_vec2 = compute_chunk_features(mp3_file)
        except:
            print "Error: Chunk Features failed"
            continue
        #title=str(track)
        title=str(dir1)+'\\'+str(track)
        print '-'*20+ title +'-'*20
        #print "       feature vector 1:",feature_vec1
        #print "       feature vector 2:",feature_vec2
        fileList.append(title)
        featureList1.append(feature_vec1)
        featureList2.append(feature_vec2)

# Write feature vecotrs of all music files to pandas data-frame
MusicFeaturesTrain=pd.DataFrame(index=fileList,data=numpy.array(featureList1),columns=FeatNames)
MusicFeaturesTrain.to_csv("FeatureFileTrainingAllList1.csv")

MusicFeaturesTest=pd.DataFrame(index=fileList,data=numpy.array(featureList2),columns=FeatNames)
MusicFeaturesTest.to_csv("FeatureFileTestAllList2.csv")

./BandCollection/Garrett/03 Clair de Lune 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/02 Who Wants to Live Forever_ 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/06 Brahms Hungarian Dance No. 5 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/15 Winter Lullaby.mp3
Error: Chunk Features failed
./BandCollection/Garrett/04 He's a Pirate (Pirates of the Car 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/08 Summer 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/14 Rock Prelude.mp3
Error: Chunk Features failed
./BandCollection/Garrett/12 New Day.mp3
Error: Chunk Features failed
./BandCollection/Garrett/10 Air.mp3
Error: Chunk Features failed
./BandCollection/Garrett/13 Ain't No Sunshine.mp3
Error: Chunk Features failed
./BandCollection/Garrett/11 Thunderstruck.mp3
Error: Chunk Features failed
./BandCollection/Garrett/01 Smooth Criminal 1.mp3
Error: Chunk Features failed
./BandCollection/Garrett/07 Chelsea Girl 1.mp3
Error: Chu

## Matching der Teilsequenzen
In diesem Abschnitt soll ein Verfahren implementiert werden, mit dem die Übereinstimmung der ersten Teilsequenz eines Musikstücks mit den zweiten Teilsequenzen aller anderen Musikstücke berechnet werden kann.



In [0]:
import math
import numpy as np
from sklearn import preprocessing
from scipy.spatial import distance

import pprint
pp = pprint.PrettyPrinter(indent=4)


**Aufagben:**
1. Lesen Sie die im vorigen Teilversuch angelegten zwei csv-Dateien in jeweils einen eigenen Pandas Dataframe ein.

In [0]:
df1 = pd.read_csv("FeatureFileTrainingAllList1.csv", sep=",", header=0, names=FeatNames)
df2 = pd.read_csv("FeatureFileTestAllList2.csv", sep=",", header=0, names=FeatNames)

2. Skalieren Sie beide Teilsequenzmengen, so dass alle Merkmale eine Standardabweichung von 1 aufweisen. Z.B. mit [http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.scale.html](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.scale.html).

In [0]:
train = preprocessing.scale(df1)
test = preprocessing.scale(df2)

Anzeigen der Mittelwerte je Datensatz ( _Training- und Testdaten_ ) und Standardabweichung.

In [13]:
pp.pprint('Train: Mean value \n {0}'.format(train.mean(axis=0)))
pp.pprint('Train: Standard deviation \n {0}'.format(train.std(axis=0)))

('Train: Mean value \n'
 ' [-0.  0.  0.  0. -0.  0. -0. -0.  0.  0.  0.  0. -0.  0.  0. -0.  0. -0.\n'
 ' -0.  0.  0.  0. -0.  0.  0.  0.  0. -0. -0. -0. -0. -0.  0. -0.  0.  0.\n'
 '  0. -0. -0.  0.  0. -0.]')
('Train: Standard deviation \n'
 ' [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n'
 ' 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]')


In [14]:
pp.pprint('Test: Mean value \n {0}'.format(test.mean(axis=0)))
pp.pprint('Test: Standard deviation \n {0}'.format(test.std(axis=0)))

('Test: Mean value \n'
 ' [ 0.  0.  0. -0.  0.  0. -0. -0. -0.  0. -0. -0. -0. -0.  0. -0. -0.  0.\n'
 '  0. -0. -0. -0. -0.  0. -0. -0. -0. -0.  0.  0.  0. -0.  0.  0. -0. -0.\n'
 '  0. -0.  0.  0. -0. -0.]')
('Test: Standard deviation \n'
 ' [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.\n'
 ' 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]')


Transformation der numpy-array in Panda Dataframes.

In [0]:
train1 = pd.DataFrame(train, index=df1.index, columns=df1.columns)
test1 = pd.DataFrame(test, index=df2.index, columns=df2.columns)

In [16]:
train1.head(5)

amp1mean   amp1std  ...    power9   power10
Adele\01 Hometown Glory.mp3        0.232196 -0.797467  ... -0.183429 -1.400576
Adele\02 I'll Be Waiting.mp3       0.180146 -0.173116  ... -0.180492 -0.854367
Adele\03 Don't You Remember.mp3    0.145859 -0.192307  ...  1.508088 -1.493972
Adele\04 Turning Tables.mp3        0.246046 -0.738488  ...  1.768903  0.295978
Adele\05 Set Fire To The Rain.mp3  0.206622 -0.324032  ...  0.410116 -0.973225

[5 rows x 42 columns]

In [17]:
test1.head(5)

amp1mean   amp1std  ...    power9   power10
Adele\01 Hometown Glory.mp3        0.252773 -0.956909  ...  0.501775 -1.077514
Adele\02 I'll Be Waiting.mp3       0.144935 -0.356562  ... -0.090862 -0.859501
Adele\03 Don't You Remember.mp3    0.178305 -0.253338  ...  1.197630 -1.512187
Adele\04 Turning Tables.mp3        0.218484 -0.722840  ...  1.185008 -0.792586
Adele\05 Set Fire To The Rain.mp3  0.151960 -0.234917  ... -0.045501 -0.780898

[5 rows x 42 columns]

Die merkmalsausprägung der Objekte unterliegt _Streuungen_. Hierdurch kann eine Distanz _d_ zwischen den Objekten durch die Mermale dominiert werden, die eine entsprechend große Streuung besitzen. Dieser Umstand ist besonders zu berücksichtigen, wenn zwischen den Objektmerkmalen, deutliche Größenunterschiede bestehen. Um die Streuung zu berücksichtigen, werden die Merkmale _skaliert_.

#### Berechnung der skalierten Distanz
**3.1 Euklidsche Distanz**: Bestimmen Sie zu jeder Teilsequenz aus der Datei _FeatureFileTrainingAllList1.csv_ die euklidische Distanz zu allen Teilsequenzen aus der Datei _FeatureFileTestAllList2.csv_ und schreiben Sie diese Distanzen in eine aufsteigend geordnete Liste. 

In [0]:
def calcEuklidDist(df1, df2):
    toRet= {}
    for index1, rows1 in df1.iterrows():
        toRet[index1] = []
        for index2, rows2 in df2.iterrows():
            toRet[index1].append([distance.euclidean(rows1, rows2), index2])
            toRet[index1].sort()
    return toRet

In [0]:
EuklidDict=calcEuklidDist(train1, test1)
#EuklidDict

**3.2 Korrelative Distanz**: Schreiben Sie auch die zugehörigen Argumente (Teilsequenzen) in eine geordnete Liste, sodass für jede Teilsequenz aus _FeatureFileTrainingAllList1.csv_ die am nächsten liegende Teilsequenz aus _FeatureFileTestAllList2.csv_ an erster Stelle steht, die zweitnächste Teilsequenz an zweiter usw.

In [0]:
def calcCorrelationDist(df1, df2):
    toRet= {}
    for index1, rows1 in df1.iterrows():
        toRet[index1] = []
        for index2, rows2 in df2.iterrows():
            toRet[index1].append([distance.correlation(rows1, rows2), index2])
            toRet[index1].sort()
    return toRet

In [0]:
CorrelationDict=calcCorrelationDist(train1, test1)
#pp.pprint(CorrelationDict)

#### Berechnung des Mittleren Rangs
4. Bestimmen Sie über alle Teilsequenzen aus _FeatureFileTrainingAllList1.csv_ den **mittleren Rang** an dem die zugehörige zweite Teilsequenz erscheint. Liegt z.B. für die erste Teilsequenz des Musikstücks A die zweite Teilsequenz nur an fünfter Stelle der geordneten nächsten Nachbarliste. Dann würde diese Teilsequenz mit dem Rang 5 in den Mittelwert einfließen.


In [0]:
from numpy import mean

def midRank(dist_dict):
    # initialize the dict
    rank_list = []
    for seq_one, seq_one_val in dist_dict.items():
        for index, seq_two_val in enumerate(seq_one_val):
            if seq_one == seq_two_val[1]:
                rank_list.append(index + 1) 
    # calculate mean
    mean_rank = mean(rank_list)
    return mean_rank

In [23]:
print("Correl mittlerer Rang:", midRank(CorrelationDict))

Correl mittlerer Rang: 2.65



5. Bestimmen Sie jetzt den mittleren Rang, für den Fall, dass _correlation_ anstelle _euclidean_ als Ähnlichkeitsmaß verwendet wird. Welches Ähnlichkeitsmaß ist für diese Anwendung zu bevorzugen?


In [24]:
print("Euklid mittlerer Rang:", midRank(EuklidDict))

Euklid mittlerer Rang: 2.5166666666666666


6. Diskutieren Sie das Ergebnis

Die euklidische Distanz dient zur Berechnung von Distanzen zwischen Vektoren (Punkte) in einem höherdimensionalen Raum. Die Korrelation ist ein Maß für den statistischen Zusammenhang zwischen zwei Datensätzen. 

Der mittlere Rang bei der Berechnung mit dem Ähnlichkeitsmaß der _correlation_ ergibt in diesem Versuch einen Wert von 2.65, während die Berechnung mit dem Ähnlichkeitsmaß der _euclidean_ einen leicht besseren Wert von 2.516 ergibt. Das Ähnlichkeitsmaß von Euklid ist in diesem Fall für die Anwendung zu bevorzugen, da dies eine bessere Aussage über die tatsächliche Ähnlichkeit der Formen zwischen den Vektoren erlaubt. 

 

## Merkmalsauswahl mit dem genetischen Algorithmus
In diesem Abschnitt soll unter Anwendung eines selbst zu implementierenden genetischen Algorithmus eine Untermenge wichtiger Merkmale aus den insgesamt 42 angelegten Merkmalen berechnet werden.
Als Vorlage kann hierfür die Implementierung für die [Lösung des TSP Problems](https://www.hdm-stuttgart.de/~maucher/Python/FunktionenAlgorithmen/html/genAlgTSP.html) herangezogen werden. Anzupassen sind dann jedoch mindestens die Fitness-Funktion, die Kreuzungs- und die Mutationsfunktion. Die Fitness soll so wie im vorigen Teilabschnitt mit dem mittleren Rang berechnet werden. Die Populationsgröße, die Anzahl der auszuwählenden Merkmale und die Anzahl der Iterationen sollen als Parameter einstellbar sein.

Der Fitnesswert des besten Individuums in der Population soll in jeder Iteration gespeichert werden. Der Verlauf dieses besten Fitness-Wertes über den Fortlauf der Iterationen soll graphisch ausgegeben werden.

Ein Pandas Frame, der nur die berechneten wichtigsten Merkmale aus _FeatureFileTrainingAllList1.csv_ enthält soll angelegt und in die csv Datei _subFeaturesTrain1.csv_ geschrieben werden.



In [0]:
from scipy.spatial import distance
from sklearn import preprocessing
import numpy as np


**Aufgaben:**
1. Implementieren Sie die die Merkmalsauswahl mit dem genetischen Algorithmus entsprechend der o.g. Beschreibung


In [0]:
def fastEuclidean(x,y):
    z=y-x
    return math.sqrt(np.dot(z,z))

Berechnung des mittleren Rangs für den Genetischen Algorithmus.

In [27]:
def mid_rank(df_train, df_test, similarity): 
    FeatureFileTrainingDF_scaled = preprocessing.scale(df_train,0)
    FeatureFileTestAllDF_scaled = preprocessing.scale(df_test,0)
    size = len(FeatureFileTestAllDF_scaled)
    rank = 0
    ranklist = np.zeros(size)
    dct_dist = np.zeros(size)
    for i,k in enumerate(FeatureFileTrainingDF_scaled):  
        for j,l in enumerate(FeatureFileTestAllDF_scaled):            
            dist = similarity(k, l)
            dct_dist[j] = dist            
        dct_index = np.argsort(dct_dist)
        ranklist[i] = np.where(dct_index == i)[0][0]+1    
    rank = sum(ranklist)/len(ranklist)
    return(rank)

2.5166666666666666

In [0]:
mid_rank(train1,test1,fastEuclidean)


## Genetischer Algorithmus für die Music Feature Selection

Die Populationsgröße, die Anzahl der auszuwählenden Merkmale und die Anzahl der Iterationen sollen als Parameter einstellbar sein.
Der Fitnesswert des besten Individuums in der Population soll in jeder Iteration gespeichert werden. Der Verlauf dieses besten Fitness-Wertes über den Fortlauf der Iterationen soll graphisch ausgegeben werden.

Ein Pandas Frame, der nur die berechneten wichtigsten Merkmale aus _FeatureFileTrainingAllList1.csv_ enthält soll angelegt und in die csv Datei _subFeaturesTrain1.csv_ geschrieben werden.

#### Pseudo Code
```
START
Generate the initial population
Compute fitness
REPEAT
    Selection
    Crossover
    Mutation
    Compute fitness
UNTIL population has converged
STOP
```

##### Generation der Population
Eine zufällige Population (_set of features_) wird aus der gesamten Population herausgelöst. Diese dient als initiale Population für den Generativen Algorithmus (GA).

In [0]:
def generate_population(popSize, featureAmount, featureNames):
    # initialize population
    population = np.zeros((popSize, featureAmount))
    # important use permutation
    for i in range(popSize):
        population[i, 0:featureAmount]=np.random.permutation(len(featureNames))[:featureAmount]
    return population

##### Implementierung des Genetischen Algorithmus


In [0]:
import numpy as np
import matplotlib.pyplot as plt

def genAlg(iterations, popSize, anzahlMerkmale, mutationPopulation, label):
    kreuzPopulation = 0.99
    
    bestDist = np.zeros(iterations)
    fitness=np.zeros(popSize)
    featureNames_fit = np.array(FeatNames)
    population = np.zeros((popSize, anzahlMerkmale))
 
    # generate initial population
    pop = generate_population(popSize, anzahlMerkmale, FeatNames)
    
    for j in range(iterations):
            #Fitnessberechnung:##############################################
            for k in range(popSize):
                selection = featureNames_fit[population[k].astype(int)]
                df1_temp = train1[selection]
                df2_temp = test1[selection]
                midRank_temp = mid_rank(df1_temp,df2_temp, fastEuclidean)
                fitness[k] = midRank_temp
            sortedIndex = fitness.argsort(axis=0)#Indizees der nach ansteigenden Distanzen sortieren
            sortedDist = fitness[sortedIndex] # die ansteigend sortiere Distanzen
            bestDist[j] = sortedDist[0] #kleinste Distanz der Iteration abspeichern    
            sortedPopulation = population[sortedIndex].astype(int) #sortierung der Population nach ansteigender Distanz    
            invertedDist = 1/sortedDist #Berechnung der Fitness aus der Distanz
            #invertedDist enthält die berechneten Fitness Werte
            #################################################################
            
            #Selection#######################################################
            invertedDistSum = invertedDist.sum()
            rn1 = invertedDistSum * np.random.rand() # Zufallszahl ziwschen 0 und 1 * invertedDistSum
            found1 = False
            index = 1
            while not found1:
                if rn1 < invertedDist[:index].sum(axis=0): #sum(axis=0): entlang der column summieren
                    found1=index
                else:
                    index+=1
            found1 = found1-1
            equal=True
            while equal:
                rn2=invertedDistSum * np.random.rand()
                found2 = False
                index=1
                while not found2:
                    if rn2 < invertedDist[:index].sum(axis=0):
                        found2 = index
                    else:
                        index+=1
                found2=found2-1
                if found2 != found1:
                    equal = False
            parent1 = sortedPopulation[found1]
            parent2 = sortedPopulation[found2]
            #parent1 und parent2 sind die selektierten Individuen
            #################################################################
            
            #Kreuzung########################################################
            crossrn = np.random.rand()
            if crossrn < kreuzPopulation:#wenn Wert innerhalb der Kreuzwahrscheinlichkeit gewürfelt -> kreuze
                
                #berechne random Index bei dem gekreuzt wird
                crossIndex = np.random.randint(0, anzahlMerkmale-1)
                
                head1, tail = np.split(parent1, [crossIndex])
                head2, tail = np.split(parent2, [crossIndex])
                
                # tail
                tailind = 0
                taillength1 = anzahlMerkmale - len(head1)
                tail1 = np.zeros(taillength1, dtype=int)
                
                for i in range(0, anzahlMerkmale):
                    if parent2[i] not in head1 and tailind < taillength1:
                        tail1[tailind] = parent2[i]
                        tailind = tailind + 1
                
                tailind = 0
                taillength2 = anzahlMerkmale - len(head2)
                tail2 = np.zeros(taillength2, dtype=int)
                for j in range(0, anzahlMerkmale):
                    if parent2[j] not in head2 and tailind < taillength2:
                        tail2[tailind] = parent2[j]
                        tailind = tailind + 1
                
                #Kind1 bekommt linken Teil von Parent1 und rechten Teil von Parent2
                child1 = np.append(head1, tail1)
                #Kind2 bekommt linken Teil von Parent2 und rechten Teil von Parent1
                child2 = np.append(head2, tail2)
            #################################################################
            
            #Mutation########################################################
            
            #Fall child1
            mutiere = np.random.rand() < mutationPopulation
            if mutiere:
                """
                wenn Wert innerhalb der Mutationswahrscheinlichkeit gewürfelt -> mutiere
                Verändere ein Merkmal des Kindes. Dabei wird das aktuelle Merkmal mit einem zufälligen Merkmal aus FeatNames
                ausgetauscht. Das neue Merkmal soll noch nicht im Kind bereits vorkommen
                """
                neuesMerkmal = np.ceil(np.random.rand()*(len(FeatNames))).astype(int)-1
                while neuesMerkmal in child1:
                    #Wenn neues Merkmal bereits im Kind enthalten, würfele neu
                    neuesMerkmal = np.ceil(np.random.rand()*(len(FeatNames))).astype(int)-1
                #wähle ein zufälliges Merkmal des Kindes aus was ersetzt wird
                altesMerkmalPos = np.ceil(np.random.rand()*anzahlMerkmale).astype(int)-1
                child1[altesMerkmalPos] = neuesMerkmal #ersetze Merkmal
            
            #Fall child2
            mutiere = np.random.rand() < mutationPopulation
            if mutiere:
                """
                wenn Wert innerhalb der Mutationswahrscheinlichkeit gewürfelt -> mutiere
                Verändere ein Merkmal des Kindes. Dabei wird das aktuelle Merkmal mit einem zufälligen Merkmal aus FeatNames
                ausgetauscht. Das neue Merkmal soll noch nicht im Kind bereits vorkommen
                """
                neuesMerkmal = np.ceil(np.random.rand()*(len(FeatNames))).astype(int)-1
                while neuesMerkmal in child2:
                    #Wenn neues Merkmal bereits im Kind enthalten, würfele neu
                    neuesMerkmal = np.ceil(np.random.rand()*(len(FeatNames))).astype(int)-1
                #wähle ein zufälliges Merkmal des Kindes aus was ersetzt wird
                altesMerkmalPos = np.ceil(np.random.rand()*anzahlMerkmale).astype(int)-1
                child2[altesMerkmalPos] = neuesMerkmal #ersetze Merkmal
            
            #child1 und child2 sind die Resultate der Mutation #######################
            
            #Ersetze die schlechtesten zwei Individuen mit den Kindern, falls die Neuen besser sind#########
            merkmaleChild1 = featureNames_fit[child1]
            df1_child1 = train1[merkmaleChild1]
            df2_child1 = test1[merkmaleChild1]
            midRank_child1 = mid_rank(df1_child1,df2_child1, fastEuclidean)
            
            merkmaleChild2 = featureNames_fit[child2]
            df1_child2 = test1[merkmaleChild2]
            df2_child2 = train1[merkmaleChild2]
            midRank_child2 = mid_rank(df1_child2,df2_child2, fastEuclidean)
            
            replace1=False
            replace2=False
            index = popSize -1
            while index > 0:
                if sortedDist[index]>midRank_child1 and not replace1:
                    if not np.ndarray.any(np.ndarray.all(child1==sortedPopulation, axis=1)):
                        sortedPopulation[index]= child1
                    replace1=True
                elif sortedDist[index]>midRank_child2 and not replace2:
                    if not np.ndarray.any(np.ndarray.all(child2==sortedPopulation, axis=1)):
                        sortedPopulation[index]= child2
                    replace2=True
                if replace1 and replace2:
                    break
                index=index-1
            population=sortedPopulation
            
    #Graphische Anzeige#########################################
    bestIndividuum = featureNames_fit[population[0]]
    print("bestIndividuum ",bestIndividuum)
    subFeaturesTrain1DF = df1[bestIndividuum]
    subFeaturesTrain1DF.to_csv('./subFeaturesTrain1.csv', sep=",")
    print("Best mid rank:", bestDist[-1])
    print("Population ", population[0])
    plt.subplot(122)
    plt.grid(True)
    plt.plot(range(iterations), bestDist, label=label)
    plt.legend()
    plt.savefig('genetic_algorithm_{}.png'.format(label))
    plt.show()
    return bestDist[-1]
    ############################################################

2. Beschreiben Sie kurz das Konzept ihrer Kreuzungs- und Mutationsfunktion. 


Kreuzungsfunktion:

Zuerst werden zwei Individuen ausgewählt. Je fitter sie sind, desto höher ist die Wahrscheinlichkeit, dass sie ausgewählt werden. Im Anschluss werden die Merkmalslisten beider Individuen an der gleichen Stelle ein zwei Teile zerschnitten. Das erste Kind bekommt dann den linken Teil von Individuum1 und den rechten Teil von Individuum2. Das zweite Kind bekommt die anderen Teile.

Mutationsfunktion:

Bei einer Mutation wird beim Kind ein zufälliges Merkmal ausgewählt und mit einem neuen Merkmal aus FeatNames, welches das Kind zuvor noch nicht besaß, ausgetauscht.


#### Training des Genetischen Algorithmus
Hierfür wird eine variierende Anzahl an Merkmalen ausgewählt, um den Algorithmus, sowie die daraus generierten Modelle auf die beste Anzahl an Merkmalen zu untersuchen. 

Während die Merkmale _iterations_, _popSize_, sowie _mutationPopulation_ für eine Vergleichbarkeit der Ergebnisse immer konstant gehalten werden, wird die Anzahl der Merkmale in immer gleich bleibenden Schritten zwischen dem Wert 10 bis 30 variiert. 

3. Bestimmen Sie eine möglichst kleine Merkmalsuntermenge mit einem möglichst guten mittleren Rang? Geben Sie sowohl die gefundenen wichtigsten Merkmale als auch den zugehörigen mittleren Rang an.


In [0]:
model_result = []

for merkmal in range(10, 22):
    m = genAlg(2000, 50, merkmal, 0.05, "{0}_features_genetic_algorithm: ".format(merkmal))
    model_result.append([merkmal, m])
    print("Distanz bei {0} Merkmalen: ".format(merkmal), m)

model_result

### Music Feature Selection

2. Implementieren und beschreiben Sie kurz das Konzept Ihrer Kreuzungs- und Mutationsfunktion.

( _Für die detailierte Implementation, siehe den Code im oberen Abschnitt._ )

##### Kreuzung:

Indivduen mit niedriger Distanz bzw. mit gutem Fitnesswert werden miteinander gekreuzt. Dabei wird ein zufälliger Schnittpunkt zur Vereinigung der Elternvektoren genutzt. Die Chromosome werden an dieser Stelle geteilt und miteinander vertauscht.

##### Mutation:

Mit niedriger Wahrscheinlichkeit wird ein Wert zufällig verändert.

3. Bestimmen Sie eine möglichst kleine Merkmalsuntermenge mit einem möglichst guten mittleren Rang? Geben Sie sowohl die gefundenen wichtigsten Merkmale als auch den zugehörigen mittleren Rang an.


Eine möglichtst kleine Merkmalsuntermenge mit möglichst gutem mittlerem Rang haben wir bei _21 Merkmalen_ gefunden. Der daraus resultierende mittlere Rang beträgt 1.41. 

Die besten daraus hervorgehenden Merkmale sind: 
```
bestIndividuum  
['amp1mean' 'amp1mean' 'power1' 'power4' 'amp1dskew' 'amp10dskew'
 'amp1000std' 'power7' 'power9' 'power8' 'amp10skew' 'amp100dstd'
 'amp100dkurt' 'power2' 'power6' 'amp1000skew' 'power10' 'amp1skew'
 'amp1mean' 'amp1mean' 'amp1mean']
```
Mit einem mittleren Rang:
```
Best mid rank: 1.4166666666666667
```
Sowie einer Population von:
```
Population  [ 0  0 32 35  6 14 25 38 40 39 10 21 23 33 37 26 41  2  0  0  0]
```


4. Um wieviel verschlechtert sich der Mittlere Rang, wenn nur die 10 wichtigsten Merkmale benutzt werden?

Ergebnis bei Anwendung von nur 10 Merkmalen:

Die besten daraus hervorgehenden Merkmale sind:
``` 
bestIndividuum  
['amp1mean' 'power1' 'power4' 'amp10kurt' 'power7' 'amp10dkurt'
 'amp100std' 'power10' 'power6' 'amp10skew']
``` 
Mit einem mittleren Rang:

```
Best mid rank: 1.6
``` 
Sowie einer Population von:
``` 
Population  [ 0 32 35 11 38 15 17 41 37 10]
``` 

Bei der Anwendung von nur 10 Merkmalen verschlechtert sich der _mittlere Rang_ um:

1.667 - 1.4333 = 0.234

Der Rang verschlechtert sich nur um 0.467. Der Unterschied zwischen 13 und 10 Merkmalen fällt bei einem mittleren Durchschnitt des Ranges von nur 1,50 groß aus. Der Informationsgewinn bei zusätzlichen Features ist somit erheblich.


#### Untersuchung der Mutation
Des weiteren wird der Algorithmus untersucht auf die Variation der Mutationswahrscheinlichkeit. Diese findet in Abständen zwischen _[0.01, 0.05, 0.1, 0.2]_ statt.

In [0]:
mutation020 = genAlg(2000, 50, 19, 0.2, "0,2")
mutation010 = genAlg(2000, 50, 19, 0.1, "0,1")
mutation005 = genAlg(2000, 50, 19, 0.05, "0,05")
mutation001 = genAlg(2000, 50, 19, 0.01, "0,01")

In [0]:
pop_mutation = [0.01, 0.05, 0.1, 0.2]

for f in pop_mutation:
    result = genAlg(2000, 50, 19, f, "{0}_features_genetic_algorithm: ".format(f))
    print("Distanz bei {0} Merkmalen: ".format(result), f)

#### Diskussion des Ergebnisses
Die Ergebnisse je Training des Genetischen Algorithmus zeigen eine deutliche Verbesserung des _mittleren Ranges_ je mehr Merkmale (engl., _Features_) zum Training verfügbar sind. Dabei ist die Verbesserungskurve besonders hoch zwischen dem Einsatz von 10 und 14 Merkmalen. Ab dem Einsatz von 21 Merkmalen (_mR_ von 1,4166) lässt sich die Verbesserungsrate des _mittleren Ranges_ nur gering verbessern.

Man kann hierdurch darauf schließen, dass bei einem größeren "genetischen Pool" einer höhere Wahrscheinlichkeit für gute Kombinationen durch Mutation und Kreuzung besteht.

## Clustering und automatische Playlistgenerierung
Implementieren Sie ein hierarchisches Clustering aller Subsequenzen in _subFeaturesTrain1.csv_. Diese _.csv_-Datei enthält nur die im vorigen Schritt ermittelten wichtigsten Merkmale. Das hierarchische Clustering ist in einem Dendrogram der Art wie in der unten gegebenen Abbildung zu visualisieren.

Die gefundenen Cluster sind mit den zugehörigen Musiktiteln in der Konsole auszugeben. 

### Das Hierarchische Clustering

Das Hierarchische Clustering ist ein Typ des unüberwachten Lernens (engl., _Unsupervised Learning_). Dieser wird angewandt, um nicht gelabelte Datenpunkte zu clustern. Ähnlich dem K-means Clustering Algorithmus grupiert der Hierarchische Clustering Algorithmus Datenpunkte zusammen, welche ähnliche Charakteristiken aufweisen. In manchen Fällen kann das Ergebnis des hierarchischen, als auch K-means Clustering Algorithmus daher ähnliche sein. Die Hierarchie dieser Cluster wird als Baum Strukture (_Dendogram_) dargestellt. Die Wurzel dieses Baumes stellt das einzigartige Cluster dar, welches alle Stichproben an Datenpunkte (engl., _samples_) vereint. In den Blättern dagegen finden sich die einzelnen Cluster, die nur einen Datenpunktsatz enthalten.



**Aufgaben:**

1. Optimieren Sie die Parameter

    1. metric (Ähnlichkeitsmaß)
    2. linkage method
    3. Clusteranzahl

Die Ausgabe des Hierarchischen Clustering Algorithmus soll wie im Folgenden angelegt sein:

![Abbildung Music Clustering](https://www.hdm-stuttgart.de/~maucher/ipnotebooks/DataMining//Bilder/playlistCluster.png "Music Clustering")

In [0]:
import sklearn
import pandas as pd
import seaborn as sns
import numpy as np
from matplotlib import pyplot as plt
from sklearn.cluster import AgglomerativeClustering
from sklearn import preprocessing

In [0]:
# Erstellung der CSV Datei mit dem besten Merkmal
genAlg(2000, 50, 21, 0.05,'21 Merkmale') 

In [0]:
subFeatureTrain = pd.read_csv("subFeaturesTrain1.csv", index_col=0)
subFeatureTrain.head(5)

#### Skalierung der Input Variablen
Um einen _Bias_ des Models zu verhindern werden die Datensätze normalisiert, sodass alle die selbe Skalierung aufweisen.

In [0]:
subFeatureTrainScaled = preprocessing.scale(subFeatureTrain)
subFeatureTrainScaled = pd.DataFrame(subFeatureTrainScaled, index=subFeatureTrain.index, columns=subFeatureTrain.columns)
subFeatureTrain = subFeatureTrainScaled

#### Datenanalyse mittels Heatmap 
Mittels einer Heatmap können die Datensätze auf deren Verteilung und Ähnlichkeit zueinander untersucht werden. Als Farbgebung für diese Heatmap wurde "Viridis" genommen. Je stärker die Felder gelb eingefärbt sind, desto größer ist der Wert dieses Merkmals. Dies verhält sich ebenso umgekehrt mit dunkel-blau eingefärbten Werten. Hierdurch lässt sich visuell analysieren, welche Merkmale einen ähnlichen Wert haben und sich damit ähnlichsten sind.

In [0]:
fig, ax = plt.subplots(figsize=(30,30))
ax.axes.set_title("Heatmap of features from music samples", fontsize=24, y=1.01)
ax.set(xlabel='Features', ylabel='Importance');
sns.heatmap(subFeatureTrain, annot=True, fmt="g", cmap='viridis', ax=ax)
plt.savefig('heatmap_music_feature_dist.png')
plt.show()

In [0]:
from scipy.cluster.hierarchy import dendrogram, linkage, fcluster

In [0]:
linked = linkage(subFeatureTrain.to_numpy(), 'complete', 'cityblock')
cluster = fcluster(linked, 6, criterion='maxclust')
labelList = subFeatureTrain.index
plt.figure(figsize=(15,15))
dendrogram(linked,
          orientation='right',
          labels=labelList,
          distance_sort='descending',
          show_leaf_counts=True,
          leaf_font_size=15,
          color_threshold=23)

for c in range(1, cluster.max()+1):
    print("Cluster", c)
    for idx,i in enumerate(cluster):
        if i == c:
            print(subFeatureTrain.index[idx])
plt.show()

1. Parameterkonstellation:

* metric (Ähnlichkeitsmaß): Cityblock

* linkage method: complete

* Clusteranzahl: 6

In [0]:
def create_dendogram(df_data, method_linkage, affinity, labels):
    #linkage matrix encoding hierarchical clustering
    Z = linkage(df_data, method_linkage, affinity) 
    plt.figure(figsize=(20, 30))
    plt.title('Hierarchical Clustering Dendrogram', fontsize=20)
    plt.xlabel('Calculated distance, method {}'.format(affinity), fontsize=16)
    plt.ylabel('Music features per title', fontsize=16)
    # create dendogram
    R = dendrogram(Z, orientation="right", labels=labels)
    # define graph spec
    ax = plt.gca()
    ax.tick_params(axis='x', which='major', labelsize=15)
    ax.tick_params(axis='y', which='major', labelsize=15)
    plt.savefig('dendrogram {}.png'.format(affinity))
    plt.show()
    
    return R

#### Unterschiedliche Affinity und Linkage
Um die verschiedenen Auswirkungen der jeweiligen Ähnlichkeitsmaße (hier: _affinity_ ), sowie Linkage Matrix (hier _linkage_ ) auf das Ergebnis des hierarchischen Clustering Plots als Dendrogram zu untersuchen werden diese im Folgenden angelegt:

In [0]:
affinity = ['euclidean', 'manhattan', 'cosine', 'precomputed']
method_linkage = ['ward', 'complete', 'average', 'single']

labels = df_music_features.music_title.tolist()
leaves_dict = []

# Create dendograms with various methods
for m in method_linkage:
  for a in affinity:
    leaves1 = create_dendogram(subFeatureTrain, m, a, labels)
    leaves_dict.append([m, a, leaves1])

Das Dictionary _R_, das von der Scipy Funktion _dendrogram()_ zurück gegeben wird, enthält Informationen über  das _Dendrogram_ und damit auch eine Liste an Werten der geclusterten Musikstüclke. 

    
2. Für welche Parameterkonstellation erlangen Sie das für Sie subjektiv betrachtet günstigste Ergebnis?

Subjektiv betrachtet ergeben die generierten Cluster (Playlists) Sinn. Künstler werden größtenteils in eine Playlist zusammengefasst, mit wenigen anderen Künstlern z.B. eine Playlist mit 15 Adele Songs und mit nur einem einzigen Lana Del Rey Song. Zudem kann man Ausreißer-Cluster anhand von kleinen Playlists erkennen.

3. Überlegen Sie sich Ansätze um diese Art der Musikgruppierung zu verbessern?

Am nahe liegendsten wäre die Extraktion von weiteren Features wie z.B. Sprache, BPM, Spieldauer der Songs. Außerdem könnte man die Clustergröße anhand von einer Mindest- oder Maximalanzahl von Songs vordefinieren, um z.B. eine Playlist mit einer gewissen Spieldauer zu erstellen. Grundsätzlich ist es möglich ein anderes Verfahren zur Clusterbildung zu verwenden (z.B. K-Means).

Fazit:

Ein Durchlauf des genetischen Algorithmus dauert sehr lange. Dies hat das Testen und Entwickeln sehr zeitaufwendig gemacht! Wir hätten diesbezüglich von Anfang an mit Multithreading arbeiten sollen, stattdessen haben wir die Performance durch eine andere Euklidfunktion verbessert.


## Zusatzaufgabe: Der Agglomerative Clustering Algorithmus

### Visuelle Analyse der Clusteranzahl mittels Dendrogram
Man kann ein Dendrogram auch zur Untersuchung der Datensätze, als auch zur Vorauswahl von Clustern heranziehen. Die Datensätze können dafür mittels einem Dendrogram näher untersucht werden und wichtige Rückschlüsse für die weitere verarbeitung gezogen werden. Dies gibt Aufschluss darüber, wie viele Cluster aus dem gesamten Datensatz (enlg., _sample_ ) im Optimum hervorgebracht werden können. Die _x-Achse_ enthält die Datensätze und die _y-Achse_ stellt den Abstand (bzw. Unähnlichkeit) zwischen den Datensätzen dar. Je größer der Abstand, desto unähnlicher sind sich die Datensätze und umgekehrt. 

In [0]:
import scipy.cluster.hierarchy as shc

Die senkrechte Linie mit dem maximalen Abstand zwischen den Sequenzen ist die blaue Linie und daher kann in diesem Beispiel ein Schwellenwert (engl., _threshold_ ) von _t=5_ und _t=3_ festgelegt werden, um hier das Dendrogramm visuell zu schneiden. Der Schwellenwert bei _t=5_ wird durch eine gestrichtelte Linie im _Dendogram_ dargestellt. Durch den Schnitt dieser mit drei zusammen laufenden Cluster Linien (engl., _fusions_) lässt sich erkennen, dass sich nun _n=3_ Cluster bilden.

In [0]:
def plot_hierarchical_cluster(data, method):
    plt.figure(figsize=(10, 7))
    plt.title("Cluster analysis Dendrogram, method={}".format(method)) 
    plt.xlabel('Music Titles Cluster')
    plt.ylabel('Euclidean distances')
    dend = shc.dendrogram(shc.linkage(data, method))
    # plot the line to make the threshhold visible
    plt.axhline(y=5, color='r', linestyle='--')
    plt.axhline(y=3, color='r', linestyle='--')
    plt.savefig('dendrogram_data_preprocessing_{}.png'.format(method))
    plt.show()

Da die Methode=_'ward'_ die besten Ergebnisse erzielt hat, wird diese auch in dem folgenden Versuch angewandt. Das Ähnlichkeitsmaß ist auch hier die _euklidsche Distanz_ .

In [0]:
plot_hierarchical_cluster(subFeatureTrain, 'ward')

#### Erkenntnis Dendrogram
Visuelle Untersuchung des Cluster Baumes in einem _Dendogram_ veranschaulicht dessen Struktur. Im untersten Knoten, dem Blatt (_engl., leave_) werden die ähnlichsten Merkmale in einem Cluster zusammen gefasst. Diese Cluster erhalten die Label zwischen _1-n_ der festgelegten Cluster Anazhl _n_. 

#### Ausreißer erkennen ( _outlier detection_ )
In dem aktuellen Graphen sind keine Ausreißer erkennbar. Diese könnten normalerweise anhand von nicht paarweise vorkommenden Zweigen (engl., _branches_ ) ausgemacht werden. 

In [0]:
def agglomerativeCluster(data, n_clusters, affinity, linkage):
    agglo_cluster_model = AgglomerativeClustering(n_clusters, affinity, linkage)
    return agglo_cluster_model.fit_predict(data)

### Trainieren des Agglomerative Clustering Algorithmus
Der Algorithmus benötigt keine initiale Angabe der Cluster Anzahl _k_, wie der k-means Clustering Algorithmus. Dies ist ein entscheidender Vorteil.

#### Vorgehen
Zunächst wird jeder Punkt als ein separates Cluster betrachtet, dann werden die Punkte rekursiv in Abhängigkeit von der Entfernung zwischen ihnen gebündelt. Die Datenpunkte werden so gruppiert, dass der Abstand zwischen den Datenpunkten innerhalb eines Clusters minimal und der Abstand zwischen den Punkten innerhalb verschiedener Clusters maximal ist. Häufig verwendete Entfernungsmaße sind die euklidische Distanz, die Manhattan-Distanz oder die Mahalanobis-Distanz. 

In [0]:
cluster_lables = agglomerativeCluster(subFeatureTrain, n_clusters=3, affinity='euclidean', linkage='ward')
cluster_lables

In [0]:
subFeatureTrain.insert((subFeatureTrain.shape[1]), 'agglomerative', cluster_lables)
subFeatureTrain.head(5)

Ausgabe der gefundenen Cluster durch den Agglomerativen Clustering Algorithmus:

In [0]:
def get_cluster(df_data, n_cluster):
    music_cluster = []
    for n in range(1, n_cluster):
        playlist_recom = df_data.loc[df_data['agglomerative'] == n]
        music_cluster.append(playlist_recom)
        print('Cluster{0}: {1}'.format(n, playlist_recom))
    return music_cluster

In [0]:
music_cluster = get_cluster(subFeatureTrain, 3)
music_cluster